### PR TITLE
Enable Nullable ReferenceTypes

### PR DESCRIPTION
**Pull request by stefannikolei**
#2231

### CODE REVIEW

File: src/ImageSharp/Metadata/Profiles/IPTC/IptcValue.cs [hunk 1]
```diff
@@ -101,7 +101,7 @@ public string Value
                 if (this.Strict && value.Length > maxLength)
                 {
                     string cappedValue = value[..maxLength];
-                    valueBytes = this.encoding.GetBytes(cappedValue);
+                    valueBytes = this.encoding!.GetBytes(cappedValue);
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Metadata/Profiles/IPTC/IptcValue.cs:104]: I think we should avoid doing this. It looks like the `Encoding` backing field should be set to `Encoding.UTF8` by default.

File: src/ImageSharp/Advanced/AdvancedImageExtensions.cs [hunk 2]
```diff
@@ -148,7 +148,7 @@ public static Memory<TPixel> DangerousGetPixelRowMemory<TPixel>(this ImageFrame<
         Guard.MustBeGreaterThanOrEqualTo(rowIndex, 0, nameof(rowIndex));
         Guard.MustBeLessThan(rowIndex, source.Height, nameof(rowIndex));
 
-        return source.PixelBuffer.GetSafeRowMemory(rowIndex);
+        return source.PixelBuffer!.GetSafeRowMemory(rowIndex);
```
Comment (stefannikolei) [inline src/ImageSharp/Advanced/AdvancedImageExtensions.cs:151]: @JimBobSquarePants Should I look for all places where I used ! and add an ArgumentNullException.ThrowIfNull?

**Comment by stefannikolei** [comment]
@JimBobSquarePants I think you could have a first look. Let me know what you think and what parts should be changed.

**Comment by tocsoft** [comment]
Hmm... there feels like there is way too much changed in this PR, this doesn't feel like the right way to tackly adding NRT into the codebase, it feel like it should be a much more targeted file by file/ section by section change, there is way to much expert knowledge required to apply some of these fixes globally, just deciding what shoudl be nullable Vs having more sensible defaults to just just changing `{get; set;}` to `{get; init;}`.

I would propose we close this huge PR (sorry about that @stefannikolei ) and instead replace it with a first PR of enable the csproj setting but inline opt out all files with warnings. This as a series of follow up PRs the various sections can be tackled.

This should me any new files/code will have it enabled by default and we can be more methodical in applying the right fixes as we go.

PS. We should also aim to tackle the public API surface first before diving into the internals.

**Comment by JimBobSquarePants** [comment]
It's with a heavy heart that I have to agree with @tocsoft here. 

I started a review of the PR and while there is some low hanging fruit - `Equals( object)` - there are several places where we are going to have to significantly refactor the codebase in order to achieve safe nullability handling. 

This is my fault. I should have investigated further before encouraging anyone to attempt a single PR.

I'm so, so sorry that you have made so much effort here with so little result.

**Comment by stefannikolei** [comment]
> It's with a heavy heart that I have to agree with @tocsoft here.
> 
> I started a review of the PR and while there is some low hanging fruit - `Equals( object)` - there are several places where we are going to have to significantly refactor the codebase in order to achieve safe nullability handling.
> 
> This is my fault. I should have investigated further before encouraging anyone to attempt a single PR.
> 
> I'm so, so sorry that you have made so much effort here with so little result.

Its ok. NRT is not an easy change. And it gets big really fast. And errors are there to be done :)

I would change it back to draft. Perhaps it is useful for anyone to look up when making the changes in smaller steps

**Comment by JimBobSquarePants** [comment]
Thanks for understanding @stefannikolei

**Comment by JimBobSquarePants** [comment]
Sorry I have broken so much! It was really important to get #2276 merged in as it was blocking development.

**Comment by stefannikolei** [comment]
@JimBobSquarePants this is the old one. I will close it. To not confuse anyone. 

I will merge your changes in my branch and fix the problems

**Pull request closed without merging**